js/editor-controller.js

```diff
@@ -0,0 +1,229 @@
+/**
+ * The Editor can be controlled by remote hosts that wrap the editor with
+ * an iframe or in some kind of webview.
+ */
+'use strict';
+
+/**
+ * @returns An object with the fs wrapper.
+ */
+var EditorController = function() {
+    var _returnObj = {};
+
+    var _editorActions = null;
+    var _controllerHost = null;
+
+    /** Constants used for iframe messaging */
+    var CONTROLLER_MESSAGING = Object.freeze({
+        // Embed editor host type
+        type: 'pyeditor',
+        // Embed editor messaging actions
+        actions: {
+            workspacesync: 'workspacesync',
+            workspacesave: 'workspacesave',
+            workspaceloaded: 'workspaceloaded',
+            importproject: 'importproject',
+            loadhex: 'loadhex',
+            loadfile: 'loadfile',
+            mobilemode: 'mobilemode',
+            savefile: 'savefile',
+            flashhex: 'flashhex',
+        }
+    });
+
+    /**
+     * JS debounce for a given amount of time. If the same function is called
+     * before the timeout the timer resets.
+     *
+     * @param {function} callback - Function to debounce. 
+     * @param {number} wait - Time to wait without calling this debounce to
+     *      finally execute the callback.
+     */
+    function debounce(callback, wait) {
+        var timeout = null;
+        return function() {
+            var args = arguments;
+            var next = function() {
+                return callback.apply(this, args);
+            };
+            clearTimeout(timeout);
+            timeout = setTimeout(next, wait);
+        }
+    }
+
+    /**
+     * Sending a postMessage to the host controller.
+     */
+    function hostPostMessage() {
+        if (_controllerHost) {
+            _controllerHost.postMessage.apply(null, arguments);
+        } else {
+            console.error('Trying to postMessage to undefined host controller:');
+            console.error(arguments);
+        }
+    }
+
```

Comment:
hostPostMessage doesn't seem to work for me. No messages arrive at the handler. `arguments` seems to be empty. The function below does deliver messages (`arguments` is still empty). What is '*' for?

```
    function hostPostMessage( msg, star) {
        console.error('hostPostMessage');
        console.error('arguments');
        console.error(arguments);
        console.error('msg');
        console.error(msg);
        console.error('star');
        console.error(star);

        if (_controllerHost) {
            _controllerHost.postMessage(msg);
        } else {
            console.error('Trying to postMessage to undefined host controller:');
            console.error( msg);
        }
     }

```

Reply:
That's weird. Does this work? I assume having the transfer value as undefined (it's not really used in any call so far) will be fine?

```js
    function hostPostMessage(message, targetOrigin, transfer) {
        if (_controllerHost) {
            _controllerHost.postMessage(message, targetOrigin, transfer);
        } else {
            console.error('Trying to postMessage to undefined host controller:');
            console.error(message, targetOrigin, transfer);
        }
    }
```

Reply:
Yes, but only message arrives in the WKScriptMessage.body NSDictionary. I don't know if we can rely on webkit postMessage quietly ignoring the extra parameters in the future.

Reply:
Yeah, you are right the `transfer` support is iffy: https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage#browser_compatibility

The targetOrigin would not really be received by the apps, but I assume the `data`, `origin`, and `source` are arriving as expected: https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage#the_dispatched_event

Reply:
Is the webkit message handler postMessage the same as window.postMessage? I have failed to find any definite documentation for the webkit one, but all references I have found use just one parameter and it's usually JSON.

https://developer.apple.com/documentation/webkit/wkusercontentcontroller/1537172-addscriptmessagehandler?language=objc

Reply:
Okay, I've pushed a commit with only 2 parameters now, hopefully this one should work correctly.

Reply:
Yes, the message parameter comes through.